client/components/domains/use-your-domain-step/index.tsx

```diff
@@ -0,0 +1,311 @@
+				</div>
+				<div className="use-your-domain-step__option-action">
+					<Button
+						className="use-your-domain-step__option-button"
+						primary={ isPrimary }
+						onClick={ onClick }
+					>
+						{ buttonText }
+					</Button>
+					<div className="use-your-domain-step__learn-more">{ learnMore }</div>
+				</div>
+			</div>
+		</Card>
+	);
+}
+
+type UseYourDomainStepProps = {
+	basePath: string;
+	goBack?: () => void;
+	initialQuery?: string;
+};
+
+function UseYourDomainStep( { basePath, goBack, initialQuery }: UseYourDomainStepProps ) {
+	const cartKey = useCartKey();
+	const { responseCart: cart } = useShoppingCart( cartKey );
+
+	const translate = useTranslate();
+
+	const [ searchQuery ] = useState( initialQuery || '' );
+
+	const selectedSite = useSelector( getSelectedSite );
+	const productsList = useSelector( getProductsList );
+	const currencyCode = useSelector( getCurrentUserCurrencyCode );
+	const domainsWithPlansOnly = useSelector( ( state ) =>
+		currentUserHasFlag( state, DOMAINS_WITH_PLANS_ONLY )
+	);
+	const primaryWithPlansOnly = useSelector( ( state ) =>
+		currentUserHasFlag( state, NON_PRIMARY_DOMAINS_TO_FREE_USERS )
+	);
+
+	const domainsWithPlansOnlyButNoPlan =
+		domainsWithPlansOnly && selectedSite?.plan && ! isPlan( selectedSite.plan );
+	const productSlug = getDomainProductSlug( searchQuery );
+	const domainProductSalePrice = getDomainTransferSalePrice(
+		productSlug,
+		productsList,
+		currencyCode
+	);
+
+	let domainProductFreeText = null;
+	if ( isNextDomainFree( cart ) || isDomainBundledWithPlan( cart, searchQuery ) ) {
+		domainProductFreeText = translate( 'Free with your plan' );
+	} else if ( domainsWithPlansOnlyButNoPlan ) {
+		domainProductFreeText = translate( 'Included in paid plans' );
+	}
+
+	let transferSalePriceText = null;
+	if (
+		! isEmpty( domainProductSalePrice ) &&
+		! isNextDomainFree( cart ) &&
+		! isDomainBundledWithPlan( cart, searchQuery ) &&
+		domainsWithPlansOnlyButNoPlan
+	) {
+		transferSalePriceText = translate( 'Sale price is %(cost)s', {
+			args: { cost: domainProductSalePrice! },
+		} );
+	}
+
+	let transferPriceText = null;
+	const domainProductPrice = getDomainPrice( productSlug, productsList, currencyCode );
+	if ( domainProductPrice ) {
+		if (
+			isNextDomainFree( cart ) ||
+			isDomainBundledWithPlan( cart, searchQuery ) ||
+			domainsWithPlansOnlyButNoPlan ||
+			getDomainTransferSalePrice( productSlug, productsList, currencyCode )
+		) {
+			transferPriceText = translate( 'Renews at %(cost)s', { args: { cost: domainProductPrice } } );
+		} else {
+			transferPriceText = translate( '%(cost)s per year', { args: { cost: domainProductPrice } } );
+		}
+	}
+
+	let mappingPriceText;
+	const price = get( productsList, [ 'domain_map', 'cost' ], null );
+	if ( price ) {
+		// @ts-expect-error despite the TS error, formatCurrency works with a
+		// `null` currencyCode and uses a fallback currency.
+		mappingPriceText = formatCurrency( price, currencyCode );
```

Comment:
I didn't want to work on `formatCurrency` to prevent scope bloat, but I get the sense, from the number of warnings that I see in the browser console, that `formatCurrency` is often invoked with a `null` or `undefined` currency code. I wonder if we should change it so that it handles `null` or `undefined` internally

Reply:
This TS error can be fixed by changing the `if ( price )` condition to `if ( price && currencyCode )`. The `currencyCode` Redux state is initialized when the products list is finished fetching (triggered by `QueryProducts`).

Reply:
> I wonder if we should change it so that it handles null or undefined internally

I think this might be hard to do. How do you format a price if you don't know the currency? Better to not render a price at all if you don't know how.

> This TS error can be fixed by changing the if ( price ) condition to if ( price && currencyCode )

I think this is the right solution.

Reply:
https://github.com/Automattic/wp-calypso/pull/101854/commits/eb60d1d141393f978699ccc03e3f9fb3a770e4ee

Reply:
One thing that could improve the `formatCurrency` situation is fixing the `getDomainPrice` function: it should return `null` if the `currencyCode` is not valid.

There are just a handful of usages, and the function already returns `null` when there is no valid "price" value, so hopefully this would be an easy refactoring.